src/Dhall/Import.hs

```diff
@@ -653,39 +690,54 @@ loadStaticWith
     -> Dhall.Core.Normalizer X
     -> Expr Src Import
     -> StateT Status m (Expr Src X)
-loadStaticWith from_import ctx n (Embed import_) = do
+loadStaticWith from_import ctx n expr₀ = case expr₀ of
+  Embed import_ -> do
     imports <- zoom stack State.get
 
-    let local (Import (ImportHashed _ (URL   {})) _) = False
-        local (Import (ImportHashed _ (Local {})) _) = True
-        local (Import (ImportHashed _ (Env   {})) _) = True
+    let local (Import (ImportHashed _ (URL     {})) _) = False
+        local (Import (ImportHashed _ (Local   {})) _) = True
+        local (Import (ImportHashed _ (Env     {})) _) = True
+        local (Import (ImportHashed _ (Missing {})) _) = True
 
     let parent = canonicalizeImport imports
     let here   = canonicalizeImport (import_:imports)
 
     if local here && not (local parent)
-        then throwM (Imported imports (ReferentiallyOpaque import_))
+        then throwMissingImport (Imported imports (ReferentiallyOpaque import_))
         else return ()
 
     expr <- if here `elem` canonicalizeAll imports
-        then throwM (Imported imports (Cycle import_))
+        then throwMissingImport (Imported imports (Cycle import_))
         else do
             m <- zoom cache State.get
             case Map.lookup here m of
                 Just expr -> return expr
                 Nothing   -> do
-                    let handler
-                            :: MonadCatch m
+                    let handler₀
```

Comment:
Is it necessary to change the logic for this handler?  My understanding is that this mainly affects how errors are rendered.  In other words, if we didn't change the handler the error might look like:

```
↳ ./foo
  ↳ Failed to resolve imports.  Error list:

      ↳ ./bar

      ↳ ./baz
```

... whereas your change causes it to render like this:

```
Failed to resolve imports. Error list:

↳ ./foo

  ↳ ./bar

↳ ./foo

  ↳ ./baz
```

... but maybe the original way might be more clear

Reply:
It is necessary (with this implementation atleast):

The reason why we need to match & unwrap the `MissingImports` here is that if we don't we will have things like `MissingImports (Imported ... (MissingImports []))` when we parse a `missing`.
This means that we _will_ render an additional "No imports found" in place of the `missing` keyword, while we wanted to avoid that.
And we do need to wrap _every_ import error in a `MissingImports`, so that the `ImportAlt` case works (so we have to do the unwrapping either here or there)

(I think my explanation might not be that clear, so I can re-explain more extensively if it doesn't make sense)

Not sure if there is a nicer way to do this whole exception piping 🤔

Reply:
Ah, I see.  Yeah, in this case keep what you have